### PR TITLE
Host verify_credentials

### PR DESCRIPTION
**Pull request by kbrock**
- parent PR: https://github.com/ManageIQ/manageiq-ui-classic/pull/8608
- depends upon: https://github.com/ManageIQ/manageiq/pull/22336

Exposes `Host#verify_credentials` for the api

This allows the user to verify credentials for a host.

/cc @MelsHyrule

### CODE REVIEW

**Comment by kbrock** [comment]
update:
- fixed spacing for hashes in 2 spec files

**Comment by MelsHyrule** [comment]
In app/controllers/api/hosts_controller.rb `class HostsController < BaseController` to `BaseProviderController`.

File: spec/requests/providers_spec.rb [hunk 1]
```diff
@@ -464,6 +464,22 @@ def have_endpoint_attributes(expected_hash)
     end
   end
 
+  describe "Providers validate" do
```
Comment (Fryguy) [inline spec/requests/providers_spec.rb:467]: The action is `verify_credentials`, so should this say:?
```suggestion
  describe "Providers verify_credentials" do
```

**Comment by kbrock** [comment]
update:
- rebased (fixes build)

File: app/controllers/api/hosts_controller.rb [hunk 2]
```diff
@@ -11,21 +12,43 @@ class HostsController < BaseController
     include Subcollections::Tags
 
     def edit_resource(type, id, data = {})
+      # TODO: drop 'credentials' parameter field when ui-classic hosts is in react
       credentials = data.delete(CREDENTIALS_ATTR)
+      authentications = data.delete(AUTH_ATTR)
       raise BadRequestError, "Cannot update non-credentials attributes of host resource" if data.any?
       resource_search(id, type).tap do |host|
-        all_credentials = Array.wrap(credentials).each_with_object({}) do |creds, hash|
+        all_credentials = Array.wrap(credentials || []).each_with_object({}) do |creds, hash|
```
Comment (agrare) [inline app/controllers/api/hosts_controller.rb:20]: is the `|| []` necessary?
```
>> Array.wrap(nil)
=> []
```

File: app/controllers/api/hosts_controller.rb [hunk 3]
```diff
@@ -11,21 +12,43 @@ class HostsController < BaseController
...
           auth_type = creds.delete(AUTH_TYPE_ATTR) || DEFAULT_AUTH_TYPE
           creds.symbolize_keys!
           creds.reverse_merge!(:userid => host.authentication_userid(auth_type))
           hash[auth_type.to_sym] = creds
         end
+        all_credentials, _ = authentications_params_to_authentications(authentications) if authentications
```
Comment (agrare) [inline app/controllers/api/hosts_controller.rb:26]: Is this overwriting `all_credentials` from up on line 20?  Hard to follow how authentications / credentials are used.

File: app/controllers/api/hosts_controller.rb [hunk 4]
```diff
@@ -11,21 +12,43 @@ class HostsController < BaseController
...
       enqueue_ems_action(type, id, "Check Compliance for", :method_name => "check_compliance", :supports => true)
     end
+
+    private
+
+    # takes credentials from params and converts into something for update_authentications
+    def authentications_params_to_authentications(authentications)
+      auth_type = authentications.keys.first
+      # symbolize userid, password
+      authentications[auth_type].symbolize_keys!
```
Comment (agrare) [inline app/controllers/api/hosts_controller.rb:49]: Should we return a new hash rather than modify the original?  If you have to modify the parameter should this be a `!` method?

File: app/controllers/api/hosts_controller.rb (hunk 3 above)
Comment (kbrock) [reply app/controllers/api/hosts_controller.rb:26]: thanks. I added comments.

The old way passed credentials.
the new way passes authentications

Yes, if they follow the new way, then we are ignoring the old way.
In theory, they will not pass the old way and the new way

Comment (kbrock) [reply app/controllers/api/hosts_controller.rb:26]: once we swap out ui classic, we can drop the old way

**Comment by kbrock** [comment]
update:
- comment around credentials vs authentications
- rename symbolize_password_keys! (twice. sorry for double push)
- fixed `Array.wrap( || [])`

File: app/controllers/api/providers_controller.rb [hunk 5]
```diff
@@ -102,15 +102,17 @@ def change_password_resource(type, id, data = {})
...
-      task_id = klass.verify_credentials_task(current_user.userid, zone_name, data)
-      action_result(true, 'Credentials sent for verification', :task_id => task_id)
-    rescue => err
-      action_result(false, err.to_s)
+    # NOTE: the ui's zone drop down is currently passing the zone_id
+    def verify_credentials_resource(type, id = nil, data = {})
+      api_action(type, id) do |klass|
+        ems_klass = fetch_provider_klass(klass, data)
+        zone_name = data.delete('zone_name')
+        zone_id = data.delete('zone_id')
```
Comment (agrare) [inline app/controllers/api/providers_controller.rb:110]: Funny I was just looking into this when I saw my `zone` was blank for verifying credentials.  This seems like its own bug that we should fix separately so we can backport what do you think?

**Comment by kbrock** [comment]
update:

- extracted provider authentication changes
- fixed issue with a pr refactor

File: app/controllers/api/providers_controller.rb (hunk 5 above)
Comment (kbrock) [reply app/controllers/api/providers_controller.rb]: see #1211

**Review by agrare — APPROVED** [review]
